drivers/usb/device/usb_dc_mcux.c

```diff
@@ -263,6 +264,14 @@ int usb_dc_ep_configure(const struct usb_dc_ep_cfg_data *const cfg)
 		return 0;
 	}
 
+	ep_addr = cfg->ep_addr;
+	status = dev_state.dev_struct.controllerInterface->deviceControl(
```

Comment:
Looks like the deinit before first init is handled solely by the fact the `dev_state` is placed in .bss section and zero initialized at startup.

Reply:
I don't quite follow - `usb_dc_attach()` and several other USB device initialization functions should be called before `usb_dc_ep_configure()` is called...

Reply:
`kUSB_DeviceControlEndpointInit` executes only within `handle_bus_reset()` (for endpoint zero) and `usb_dc_ep_configure()`. Inside `usb_dc_ep_configure()` the `kUSB_DeviceControlEndpointDeinit` is called before `kUSB_DeviceControlEndpointInit`, therefore on first call to `usb_dc_ep_configure()` the endpoint is first deinitialized (before it was ever initialized), and then initialized. It works because the `dev_state` is filled with zeroes on system startup and `kUSB_DeviceControlEndpointDeinit` simplky returns if the allocated buffer is outside the range.

Reply:
OK - but if you look at the call:
```
	status = dev_state.dev_struct.controllerInterface->deviceControl(
			dev_state.dev_struct.controllerHandle,
			kUSB_DeviceControlEndpointDeinit, &ep_addr);
```
only `dev_state.dev_struct.controllerHandle` from `dev_state` is passed into the `deviceControl()` function for the HAL. This `dev_state.dev_struct.controllerHandle` is initialized during `usb_dc_attach()`. Also, it is implicitly assumed that `dev_state.dev_struct.controllerInterface` is properly initialized for the device being used. The particular device's HAL also has some state which needs some bare minimum initialization (this state is private to the HAL and separate from `dev_state`).

But you do bring up a good point that this this approach to always `Deinit` the endpoint before `Init` ought to really be tested on all devices covered by the the MCUX HAL. As I only have one EVK in my hands, I could use some help with testing...